xapian-core/weight/tfidfweight.cc

```diff
@@ -36,54 +36,110 @@ using namespace std;
 
 namespace Xapian {
 
-TfIdfWeight::TfIdfWeight(const std::string &normals)
-    : normalizations(normals), param_slope(0.2), param_delta(1.0)
+TfIdfWeight::TfIdfWeight(const std::string& normals)
+    : TfIdfWeight::TfIdfWeight(normals, 0.2, 1.0) {}
+
+TfIdfWeight::TfIdfWeight(const std::string& normals, double slope, double delta)
+    : param_slope(slope), param_delta(delta)
 {
-    if (normalizations.length() != 3 ||
-	!strchr("nbslPL", normalizations[0]) ||
-	!strchr("ntpfsP", normalizations[1]) ||
-	!strchr("n", normalizations[2]))
+    if (normals.length() != 3 ||
+	!strchr("nbslPL", normals[0]) ||
+	!strchr("ntpfsP", normals[1]) ||
+	!strchr("n", normals[2]))
 	throw Xapian::InvalidArgumentError("Normalization string is invalid");
-    if (normalizations[1] != 'n') {
+    if (param_slope <= 0)
+	throw Xapian::InvalidArgumentError("Parameter slope is invalid");
+    if (param_delta <= 0)
+	throw Xapian::InvalidArgumentError("Parameter delta is invalid");
+    if (normals[1] != 'n') {
 	need_stat(TERMFREQ);
 	need_stat(COLLECTION_SIZE);
     }
     need_stat(WDF);
     need_stat(WDF_MAX);
     need_stat(WQF);
-    if (normalizations[0] == 'L') {
+    if (normals[0] == 'P' || normals[1] == 'P') {
+	need_stat(AVERAGE_LENGTH);
+	need_stat(DOC_LENGTH);
+	need_stat(DOC_LENGTH_MIN);
+    }
+    if (normals[0] == 'L') {
 	need_stat(DOC_LENGTH);
 	need_stat(DOC_LENGTH_MIN);
 	need_stat(DOC_LENGTH_MAX);
 	need_stat(UNIQUE_TERMS);
     }
+    switch (normals[0]) {
+	case 'b':
+	    wdf_norm_ = wdf_norm::BOOLEAN;
+	    break;
+	case 's':
+	    wdf_norm_ = wdf_norm::SQUARE;
+	    break;
+	case 'l':
+	    wdf_norm_ = wdf_norm::LOG;
+	    break;
+	case 'P':
+	    wdf_norm_ = wdf_norm::PIVOTED;
+	    break;
+	case 'L':
+	    wdf_norm_ = wdf_norm::LOG_AVERAGE;
+	    break;
+	default:
+	    wdf_norm_ = wdf_norm::NONE;
+    }
+    switch (normals[1]) {
+	case 'n':
+	    idf_norm_ = idf_norm::NONE;
+	    break;
+	case 's':
+	    idf_norm_ = idf_norm::SQUARE;
+	    break;
+	case 'f':
+	    idf_norm_ = idf_norm::FREQ;
+	    break;
+	case 'P':
+	    idf_norm_ = idf_norm::PIVOTED;
+	    break;
+	case 'p':
+	    idf_norm_ = idf_norm::PROB;
+	    break;
+	default:
+	    idf_norm_ = idf_norm::TFIDF;
+    }
+    wt_norm_ = wt_norm::NONE;
 }
 
-TfIdfWeight::TfIdfWeight(const std::string &normals, double slope, double delta)
-    : normalizations(normals), param_slope(slope), param_delta(delta)
+TfIdfWeight::TfIdfWeight(wdf_norm wdf_normalization,
+			 idf_norm idf_normalization,
+			 wt_norm wt_normalization)
+    : TfIdfWeight::TfIdfWeight(wdf_normalization, idf_normalization,
+			       wt_normalization, 0.2, 1.0) {}
+
+TfIdfWeight::TfIdfWeight(wdf_norm wdf_normalization,
+			 idf_norm idf_normalization,
+			 wt_norm wt_normalization,
+			 double slope, double delta)
+    : wdf_norm_(wdf_normalization), idf_norm_(idf_normalization),
+      wt_norm_(wt_normalization), param_slope(slope), param_delta(delta)
 {
```

Comment:
This and the constructor which takes a 3 character code for compatibility share all the logic to select which stats are needed, and to vet the other parameters.

It'd be much cleaner for the compatibility one to forward to this, then we only need to have one copy of that code.

You just need to arrange that the decoding code can be done in the forwarding constructor call, e.g. with a function for each one, something like:

```
TfIdfWeight::TfIdfWeight(const string& normals, double slope, double delta)
    : TfIdfWeight(decode_wdf_norm(normals).
                  decode_idf_norm(normals),
                  decode_wt_norm(normals),
                  slope,
                  delta) {}
```

Reply:
Updated